src/k2/httpproxy/HTTPProxy.cpp

```diff
@@ -209,12 +218,12 @@ seastar::future<nlohmann::json> HTTPProxy::_handleEnd(nlohmann::json&& request)
         return JsonResponse(Statuses::S400_Bad_Request("Bad json for end request"));
     }
 
-    std::unordered_map<uint64_t, k2::K2TxnHandle>::iterator it = _txns.find(id);
-    if (it == _txns.end()) {
+    auto iter = _txns.find(id, false);
+    if (iter == _txns.end()) {
         return JsonResponse(Statuses::S400_Bad_Request("Could not find txnID for end request"));
     }
 
-    return it->second.end(commit)
+    return _txns.getValue(iter).txn.end(commit)
     .then([this, id] (k2::EndResult&& result) {
         _txns.erase(id);
```

Comment:
I think this is a race condition with the timeout timer expiration. You could have two in flight end() calls, which is ok, but the txn will be erased twice. If you make the erase call here safe to call on a non-existing item then it should be fine.

Reply:
Good catch! Though the operations seem to run in the same thread, they can interleave? Added more check in the `erase` function. The other place it can happen is `at(...)` function which will throw exception. Leaving it as is, as it's a rare condition and will at worst result in a 500 error caused by uncaught exception instead of 400 error.

Reply:
It is in the same thread but can be concurrent because of the Seastar execution model. The call to end() here ends up as a remote async call. Then the timer could fire and be executed, and after that the .then block here is executed. I think you still need to fix it in the wrapper class because the underlying call to the unordered_map erase with an end iterator is undefined, not an exception.

Reply:
Yes, I added a check inside erase function in the previous update but that got mistakenly reverted. I have put back the check.